### PR TITLE
BUG: ENH: Check if guess for newton is already zero before checking derivative

### LINKED ISSUE #8904: BUG: if zero derivative at root, then Newton fails with RuntimeWarning
As indicated in [this discussion](../pull/8357#discussion_r192612946) by @pvanmulbregt

>... if the derivative is 0 at the root? E.g. `f(x)=x**3 - x**2`, starting at `x0=0` (immediately at the root), ... It looks like this might be considered a failure

### Reproducing code example:
```
from scipy import optimize

# use secant
r = optimize.newton(lambda x: x**3 - x**2, x0=0)
# -1.3552527156068808e-20  close enough to zero to be the correct answer

# now try Newton
r = optimize.newton(lambda x: x**3 - x**2, x0=0, fprime=lambda x: 3*x**2 - 2*x)
# raises RuntimeWarning but solution is zero

# or Halley's
r = optimize.newton(lambda x: x**3 - x**2, x0=0, fprime=lambda x: 3*x**2 - 2*x, fprime2=lambda x: 6*x - 2)
# also raises RuntimeWarning even though solution should be zero
```

### Error message:
```
~/miniconda/lib/python3.6/site-packages/scipy/optimize/zeros.py:169: RuntimeWarning: derivative was zero.
  warnings.warn(msg, RuntimeWarning)
```

### Scipy/Numpy/Python version information:
```
1.1.0 1.14.3 sys.version_info(major=3, minor=6, micro=5, releaselevel='final', serial=0)
```

### PR DESCRIPTION
**Pull request by mikofski**
* fixes #8904 
* adds a test to see if `RuntimeWarning` is raised for case where `fval(p0)` is already zero where `p0` is the initial guess
* checks in `newton` if `fval(p0)` is already zero, before iterating for any method or evaluating derivatives `fprime` (or `fprime2`) if given, and returns `_ECONVERGED` instead of raising a `RuntimeWarning`
* fails the current test `scipy/optimize/tests/test_zeros.py::TestBasic::test_deriv_zero_warning` since these two are in opposition.
* see [post on SciPy Dev mailing list](https://mail.python.org/pipermail/scipy-dev/2018-June/022861.html) for more info

### CODE REVIEW

File: scipy/optimize/zeros.py [hunk 1]
```diff
@@ -187,6 +187,9 @@ def newton(func, x0, fprime=None, args=(), tol=1.48e-8, maxiter=50,
     # Multiply by 1.0 to convert to floating point.  We don't use float(x0)
     # so it still works if x0 is complex.
     p0 = 1.0 * x0
+    # check if initial guess is the solution
+    if func(p0, *args) == 0:
+        return _results_select(full_output, (p0, 0, 0, _ECONVERGED))
```
Comment (pvanmulbregt) [inline scipy/optimize/zeros.py]: The check outside the loop only handles the case of the starting point being a root.  Its doesn't handle an intermediate iterate being a multiple root.
Move the test inside the loop, and move the evaluation of `f` ahead of that of `fprime`.  Something like
```
        for itr in range(maxiter):
            # First evaluate f
            fval = func(p0, *args)
            funcalls += 1
            # If fval is 0, a root has been found. Terminate.
            if fval == 0:
                return _results_select(full_output, (p0, funcalls, itr, _ECONVERGED))
            # Now evaluate f'
            fder = fprime(p0, *args)
            funcalls += 1
            # If fder is 0 and fval != 0, there is a problem. Warn/terminate
            if fder == 0:
                msg = "derivative was zero."
                warnings.warn(msg, RuntimeWarning)
                return _results_select(full_output, (p0, funcalls, itr + 1, _ECONVERR))
            newton_step = fval / fder
           ...

```

**Review by pvanmulbregt — CHANGES_REQUESTED** [review]

**Comment by mikofski** [comment]
So then is there _anyone_ who is **not** okay with changing the existing tests?

File: scipy/optimize/zeros.py (hunk 1 above)
Comment (mikofski) [reply scipy/optimize/zeros.py]: I agree, in that case shall I change the title of the PR to "check for convergence using abs of residual" or something like that?

And more importantly, what tolerance should we use to check for convergence in this way? I think a test for zero identically is too strict, perhaps something more like:

```python
# If a root has been found within some tolerance, then terminate
    if abs(fval) < tol:
                return _results_select(full_output, (p0, funcalls, itr, _ECONVERGED))
```

Should we use the same tolerance as what's currently used for the Newton step or add a new argument for a separate one? Or should we just check if the guess is identically zero?

Comment (pvanmulbregt) [reply scipy/optimize/zeros.py]: Testing for "exactly 0" fixes the situation of landing in a multiple root.
Testing for "approximately 0" would change existing results.  It also isn't scale-invariant (replacing `f` with `const * f` leads to different outcomes when using "approximately 0".).
So "exactly 0" is the right criteria to apply.

**Comment by mikofski** [comment]
So @pvanmulbregt, is your review affirmative now? @rgommers I believe this fixes the bug #8904 and IMO is an improvement to the existing SciPy NR implementation. Let me know if you have any other comments. Thanks!

**Review by pvanmulbregt — APPROVED** [review]

File: scipy/optimize/tests/test_zeros.py [hunk 2]
```diff
@@ -169,3 +169,36 @@ def f_2(x, *a):
     y = zeros.newton(f, z, args=coeffs, fprime=f_1, fprime2=f_2, tol=1e-6)
     # (-0.75000000000000078+1.1989578808281789j)
     assert_allclose(f(y, *coeffs), 0, atol=1e-6)
+
+
+@pytest.mark.filterwarnings("error")
```
Comment (rgommers) [inline scipy/optimize/tests/test_zeros.py]: this line can be removed, this is the default behavior (see `pytest.ini` line 7 in the repo root).

File: scipy/optimize/tests/test_zeros.py [hunk 3]
```diff
@@ -169,3 +169,36 @@ def f_2(x, *a):
...
+def test_gh8904_zeroder_at_root_fails():
+    """Test that Newton or Halley don't warn if zero derivative at root"""
+
+    # a function that has a zero derivative at it's root
+    def f_zeroder_root(x):
+        return x ** 3 - x ** 2
```
Comment (rgommers) [inline scipy/optimize/tests/test_zeros.py]: formatting: there should be no spaces around `**` (here and below)

File: scipy/optimize/tests/test_zeros.py [hunk 4]
```diff
@@ -169,3 +169,36 @@ def f_2(x, *a):
...
+
+    # 1st derivative
+    def fder(x):
+        return 3 * x ** 2 - 2 * x
+
+    # should work with newton and halley
+    r = zeros.newton(f_zeroder_root, x0=0, fprime=fder)
+    assert_allclose(r, 0, atol=zeros._xtol, rtol=zeros._rtol)
+    r = zeros.newton(f_zeroder_root, x0=0, fprime=fder,
+                     fprime2=lambda x: 6 * x - 2)
```
Comment (rgommers) [inline scipy/optimize/tests/test_zeros.py]: formatting nitpick: `6*x - 2`

**Comment by rgommers** [comment]
Thanks @mikofski. I've made a few minor comments, will be happy to merge once those are addressed.

File: scipy/optimize/tests/test_zeros.py (hunk 2 above)
Comment (mikofski) [reply scipy/optimize/tests/test_zeros.py]: Hi @rgommers, even though the `error` warning filter is set by default, it didn't seem to work when I use `python runtest.py -t scipy/optimize/tests/test_zeros.py` and more importantly I feel that it's more explicit to leave the decorator even if it is redundant.

Comment (rgommers) [reply scipy/optimize/tests/test_zeros.py]: We want to control this globally. For one, we're switching this behavior between development versions and releases (we do want errors for the former, not for the latter).

Comment (mikofski) [reply scipy/optimize/tests/test_zeros.py]: Okay

**Comment by mikofski** [comment]
Thanks for your comments @rgommers - hopefully all good now.

**Comment by rgommers** [comment]
Let's give this a go. Thanks @mikofski!

**Pull request merged**